contracts/protocol/bases/PriceDiscoveryBase.sol

```diff
@@ -62,124 +72,250 @@ contract PriceDiscoveryBase is ProtocolBase {
+
+            // Check the native balance and return the surplus to seller
+            uint256 protocolNativeBalanceAfter = getBalance(address(0), owner);
+
+            if (protocolNativeBalanceAfter > protocolNativeBalanceBefore) {
+                // Return the surplus to seller
+                FundsLib.transferFundsFromProtocol(
+                    address(0),
+                    payable(msgSender()),
+                    protocolNativeBalanceAfter - protocolNativeBalanceBefore
+                );
+            }
+        }
+
+        verifyStateAfterPriceDiscoveryCall(bosonVoucher, _tokenId);
 
+        clearStorage(_tokenId);
+    }
+
+    /**
+     * @notice Fulfils a bid order on external contract when caller is voucher owner.
+     * Protocol act on behalf of the voucher owner
+     *
+     * Reverts if:
+     * - Offer price is in native token and caller does not approve WETH
+     * - Calling transferFrom on token fails for some reason (e.g. protocol is not approved to transfer)
+     * - Received ERC20 token amount differs from the expected value
+     * - Transfer of voucher to the buyer fails for some reason (e.g. buyer is contract that doesn't accept voucher)
+     * - Call to price discovery contract fails
+     * - Exchange doesn't exist after the call to price discovery contract
+     * - Exchange is not in the committed state
+     * - Voucher owner did not approve protocol to transfer the voucher
+     * - Price received from price discovery is lower than the expected price
+     * - Reseller did not approve protocol to transfer exchange token in escrow
+     * - New voucher owner is not buyer wallet
+     *
+     * @param _tokenId - the id of the token
+     * @param _offer - the fully populated BosonTypes.Offer struct
+     * @param _priceDiscovery - the fully populated BosonTypes.PriceDiscovery struct
+     * @return actualPrice - the actual price of the order
+     */
+    function fulfilBidCallerIsOwner(
+        uint256 _tokenId,
+        Offer storage _offer,
+        PriceDiscovery calldata _priceDiscovery,
+        ProtocolLib.ProtocolLookups storage _lookups,
+        IBosonVoucher bosonVoucher
+    ) internal returns (uint256 actualPrice) {
         // Transfer seller's voucher to protocol
         // Don't need to use safe transfer from, since that protocol can handle the voucher
-        bosonVoucher.transferFrom(msgSender(), address(this), _exchangeId);
+        bosonVoucher.transferFrom(msgSender(), address(this), _tokenId);
+
+        // Owner is now protocol
+        address owner = address(this);
+
+        // Approve price discovery contract to transfer voucher. There is no need to reset approval afterwards, since protocol is not the voucher owner anymore
+        bosonVoucher.approve(_priceDiscovery.priceDiscoveryContract, _tokenId);
 
-        if (_exchangeToken == address(0)) _exchangeToken = address(weth);
+        address exchangeToken = _offer.exchangeToken;
```

Comment:
Some PD don't support native tokens for bid side (seaPort for example). On the other hand some don't automatically wrap ether (LSSVM for example), so we need to decide how to handle both cases.

Reply:
LSSVM works with both native currency and erc20
Is not safe to use native token as the last voucher owner can be a contract with a receive function that intercepts any funds send to it
 So I guess we should always use WETH